### PR TITLE
minimize impact of invalidations due to uninferred `ctx`

### PR DESCRIPTION
**Pull request by thchr**
I originally tried to fix this in SHA.jl, but it is more probably more appropriate to fix it here. The problem is summarized here: https://github.com/JuliaCrypto/SHA.jl/pull/88#issuecomment-1516151954

### CODE REVIEW

File: src/extract.jl [hunk 1]
```diff
@@ -285,7 +285,15 @@ function git_object_hash(
     kind::AbstractString,
     HashType::DataType,
```
Comment (StefanKarpinski) [inline src/extract.jl:286]: I think you can just change the signature to `::Type{HashType} …. where {HashType}` instead, no?

Comment (thchr) [reply src/extract.jl:286]: That might be better then, yeah.
I could also just instantiate `HashType` to a `context` at the outset?
I thought maybe the lack of specialization was intentional.

Comment (StefanKarpinski) [reply src/extract.jl:286]: The issue with the latter approach is that you need many context instances while computing git-style tree hashes, not just one context object.

Comment (thchr) [reply src/extract.jl]: Makes sense, thanks.

I changed to `::Type{HashType}, ...) where HashType` now. I also changed the branching for calling `git_tree_hash` to hopefully eliminate any other instabilities at their origin.

**Comment by thchr** [comment]
Gentle bump :)

**Review by StefanKarpinski — APPROVED** [review]

**Pull request merged**